storage/watchable_store_test.go

```diff
@@ -59,6 +59,139 @@ func TestNewWatcherCancel(t *testing.T) {
 	}
 }
 
+func TestCancelUnsynced(t *testing.T) {
+	// manually create watchableStore instead of newWatchableStore
+	// because newWatchableStore automatically calls syncWatchers
+	// method to sync watchers in unsynced map. We want to keep watchers
+	// in unsynced to test if syncWatchers works as expected.
+	s := &watchableStore{
+		store:    newStore(tmpPath),
+		unsynced: make(map[*watching]struct{}),
+
+		// to make the test not crash from assigning to nil map.
+		// 'synced' doesn't get populated in this test.
+		synced: make(map[string]map[*watching]struct{}),
+	}
+
+	defer func() {
+		s.store.Close()
+		os.Remove(tmpPath)
+	}()
+
+	// Put a key so that we can spawn watchers on that key.
+	// (testKey in this test). This increases the rev to 1,
+	// and later we can we set the watcher's startRev to 1,
+	// and force watchers to be in unsynced.
+	testKey := []byte("foo")
+	testValue := []byte("bar")
+	s.Put(testKey, testValue)
+
+	w := s.NewWatcher()
+
+	// arbitrary number for watcher
+	watcherSize := 100
+
+	cancels := make([]CancelFunc, watcherSize)
+	for i := 0; i < watcherSize; i++ {
+		// non-0 value to keep watchers in unsynced
+		_, cancel := w.Watch(testKey, true, 1)
+		cancels[i] = cancel
+	}
+
+	// Before running CancelFunc
+	//
+	// synced should be empty
+	// because we manually populate unsynced only
+	if len(s.synced[string(testKey)]) != 0 {
+		t.Errorf("synced[string(testKey)] size = %d, want 0", len(s.synced[string(testKey)]))
```

Comment:
actually i think we should not repeat this check here. we already check this behavior in `TestSyncWatchings`. For this test, we only wants to ensure the canceling behavior.